### PR TITLE
add patch endpoint for jobs and etags

### PR DESCRIPTION
**Pull request by justinpjose**
### What
**Describe what you have changed and why.**
- Add eTag functionality
- Add patch job status handler
- Add unit test for patch endpoint

### How to review
**Describe the steps required to test the changes.**
- Check if the code changes make sense
- Check if the tests pass

### Who can review
**Describe who worked on the changes, so that other people can review.**
- Anyone

### CODE REVIEW

File: api/jobs.go [hunk 1]
```diff
@@ -297,3 +307,211 @@ func (api *API) updateSearchIndexName(ctx context.Context, reindexResponse *http
...
+
+	// set eTag on ETag response header
+	dpresponse.SetETag(w, newETag)
+
+	w.WriteHeader(http.StatusNoContent)
+
+	log.Info(ctx, "successfully patched status of job", logData)
+}
+
+// getUpdatesFromJobPatches returns an updated job resource and updated bson.M resource based on updates from the patches
```
Comment (cookel2) [inline api/jobs.go:431]:
```suggestion
// GetUpdatesFromJobPatches returns an updated job resource and updated bson.M resource based on updates from the patches
```

File: api/tasks_test.go [hunk 2]
```diff
@@ -36,6 +36,32 @@ const (
 	bindAddress           = "localhost:25700"
 )
 
+// Create Task Payload
+var createTaskPayloadFmt = `{
+	"task_name": "%s",
+	"number_of_documents": 5
+}`
+
+func expectedTask(jobID string, lastUpdated time.Time, numberOfDocuments int, taskName string) (models.Task, error) {
```
Comment (cookel2) [inline api/tasks_test.go:45]: GoDoc needed

File: api/jobs_test.go [hunk 3]
```diff
@@ -681,3 +698,481 @@ func TestPutNumTasksHandler(t *testing.T) {
...
+				Op:    dprequest.OpReplace.String(),
+				Path:  "/unknown",
+				Value: "unknown",
+			},
+		}
+
+		Convey("When preparePatchUpdates is called", func() {
+			updatedJob, bsonUpdates, err := api.GetUpdatesFromJobPatches(testCtx, unknownPathPatches, currentJob, log.Data{})
+
+			Convey("And an error should be returned", func() {
```
Comment (cookel2) [inline api/jobs_test.go:1081]:
```suggestion
			Convey("Then an error should be returned", func() {
```

File: api/jobs_test.go [hunk 4]
```diff
@@ -681,3 +698,481 @@ func TestPutNumTasksHandler(t *testing.T) {
...
+				Op:    dprequest.OpReplace.String(),
+				Path:  models.JobNoOfTasksPath,
+				Value: "unknown",
+			},
+		}
+
+		Convey("When preparePatchUpdates is called", func() {
+			updatedJob, bsonUpdates, err := api.GetUpdatesFromJobPatches(testCtx, invalidNoOfTasksPatches, currentJob, log.Data{})
+
+			Convey("And an error should be returned", func() {
```
Comment (cookel2) [inline api/jobs_test.go:1103]:
```suggestion
			Convey("Then an error should be returned", func() {
```
Please also change the ones below this to be 'Then'.

File: api/jobs_test.go [hunk 5]
```diff
@@ -681,3 +698,481 @@ func TestPutNumTasksHandler(t *testing.T) {
...
+				Op:    dprequest.OpReplace.String(),
+				Path:  models.JobStatePath,
+				Value: "unknown",
+			},
+		}
+
+		Convey("When preparePatchUpdates is called", func() {
+			updatedJob, bsonUpdates, err := api.GetUpdatesFromJobPatches(testCtx, unknownStatePatches, currentJob, log.Data{})
+
+			Convey("And an error should be returned", func() {
```
Comment (cookel2) [inline api/jobs_test.go:1125]:
```suggestion
			Convey("Then an error should be returned", func() {
```

**Review by cookel2 — CHANGES_REQUESTED** [review]
Justin I've only made a few comments but I see that you have made some great improvements to the whole repo and sorted out the mocks, which I had mistakenly modified, too.

File: api/jobs.go (hunk 1 above)
Comment (justinpjose) [reply api/jobs.go]: Good spot

File: api/jobs_test.go (hunk 3 above)
Comment (justinpjose) [reply api/jobs_test.go]: Done

File: api/jobs_test.go (hunk 5 above)
Comment (justinpjose) [reply api/jobs_test.go]: Done

File: api/jobs_test.go [hunk 6]
```diff
@@ -690,3 +698,481 @@ func TestPutNumTasksHandler(t *testing.T) {
...
+			{
+				Op:   dprequest.OpRemove.String(),
+				Path: models.JobTotalSearchDocumentsPath,
+			},
+		}
+
+		Convey("When preparePatchUpdates is called", func() {
+			updatedJob, bsonUpdates, err := api.GetUpdatesFromJobPatches(testCtx, unknownOpPatches, currentJob, log.Data{})
+
+			Convey("And an error should be returned", func() {
```
Comment (cookel2) [inline api/jobs_test.go:1059]:
```suggestion
			Convey("Then an error should be returned", func() {
```

File: api/jobs_test.go [hunk 7]
```diff
@@ -690,3 +698,481 @@ func TestPutNumTasksHandler(t *testing.T) {
...
+				Op:    dprequest.OpReplace.String(),
+				Path:  models.JobStatePath,
+				Value: 12,
+			},
+		}
+
+		Convey("When preparePatchUpdates is called", func() {
+			updatedJob, bsonUpdates, err := api.GetUpdatesFromJobPatches(testCtx, invalidStatePatches, currentJob, log.Data{})
+
+			Convey("And an error should be returned", func() {
```
Comment (cookel2) [inline api/jobs_test.go:1147]:
```suggestion
			Convey("Then an error should be returned", func() {
```

File: api/jobs_test.go [hunk 8]
```diff
@@ -690,3 +698,481 @@ func TestPutNumTasksHandler(t *testing.T) {
...
+				Op:    dprequest.OpReplace.String(),
+				Path:  models.JobTotalSearchDocumentsPath,
+				Value: "invalid",
+			},
+		}
+
+		Convey("When preparePatchUpdates is called", func() {
+			updatedJob, bsonUpdates, err := api.GetUpdatesFromJobPatches(testCtx, invalidTotalSearchDocsPatches, currentJob, log.Data{})
+
+			Convey("And an error should be returned", func() {
```
Comment (cookel2) [inline api/jobs_test.go:1169]:
```suggestion
			Convey("Then an error should be returned", func() {
```

**Review by cookel2 — CHANGES_REQUESTED** [review]
Justin I just found a few more places where 'And' needs to be changed to 'Then'.

File: api/jobs_test.go (hunk 6 above)
Comment (justinpjose) [reply api/jobs_test.go]: Sorry about that. Thank you for spotting it

File: api/jobs_test.go (hunk 7 above)
Comment (justinpjose) [reply api/jobs_test.go]: Done

File: api/jobs_test.go (hunk 8 above)
Comment (justinpjose) [reply api/jobs_test.go]: Done

**Pull request merged**